Add toggle for inserting keys/markers at current time vs mouse cursor's position

shadow-foss opened the pull request:
## Summary
Adds a toggle button in the Animation Track Editor that allows users to choose between inserting animation keys and markers at the timeline cursor position or at the mouse position.

## Changes
- Add `editors/animation/insert_at_timeline_cursor` editor setting to persist toggle state
- Add toggle button UI in Animation Track Editor toolbar  
- Modify key/marker insertion logic to use cursor position when toggle is enabled
- Add documentation for the new editor setting
- Icon by @TokageItLab

Fixes #103272

## Demo

https://github.com/user-attachments/assets/770923d0-2976-4769-8126-3945efcd8d63

_Bugsquad edited:_
- Supersedes/closes https://github.com/godotengine/godot/pull/105774

shadow-foss committed:
Add toggle to insert keys/markers at current time or mouse position

Adds a new editor setting editors/animation/insert_at_current_time and a toggle button in the Animation Track Editor to let users choose whether to insert keys and markers at the current timeline cursor (when enabled) or at the mouse position (default behavior).

- Key insertion
- Paste and duplicate operations
- Editor setting persistence
- Icon by @TokageItLab

Fixes #103272

shadow-foss commented:
Quick question: the toggle is enabled by default, so keys/markers go to the timeline cursor.
Previously, they were added at the mouse cursor by default.
Could this change be confusing?
Should I set the default to disabled for consistency?

Happy to change it if needed!

SatLess commented:
Imo I think this should be an opt in, so disabled by default

TokageItLab commented on doc/classes/EditorSettings.xml:
```diff
@@ -547,6 +547,9 @@
 		<member name="editors/animation/default_fps_mode" type="int" setter="" getter="">
 			Default step mode for [AnimationPlayer] (seconds or FPS). The option is remembered locally for a scene and this option only determines the default value when scene doesn't have local state yet.
 		</member>
+		<member name="editors/animation/insert_at_timeline_cursor" type="bool" setter="" getter="">
```
```suggestion
		<member name="editors/animation/insert_at_current_time" type="bool" setter="" getter="">
```

Since the name “timeline cursor” is not used elsewhere, I feel that the word “cursor” is a little confusing.

Mickeon replied:
It's referred as such in the "Apply snapping to timeline cursor." tooltip. Most of the times it's actually just called "cursor", and there's precedence with the `move_last_selected_key_to_cursor` action, so I would name it `insert_at_cursor`, maybe.

TokageItLab replied:
> Apply snapping to timeline cursor.

Oh indeed, there is the word.

> You can see why OP decided to name it this way. It's ambiguous with the mouse cursor.

But well, that's what I want to say. I assume "at timeline cursor" can be confused "on mouse cursor on the timeline", so I recommend changing the options and variable names in conjunction with the [suggestion](https://github.com/godotengine/godot/pull/107511#discussion_r2146749028) below.

TokageItLab commented on editor/editor_settings.cpp:
```diff
@@ -959,6 +959,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("editors/animation/confirm_insert_track", true, true);
 	_initial_set("editors/animation/default_create_bezier_tracks", false, true);
 	_initial_set("editors/animation/default_create_reset_tracks", true, true);
+	_initial_set("editors/animation/insert_at_timeline_cursor", true, true);
```
```suggestion
	_initial_set("editors/animation/insert_at_timeline_cursor", false, true);
```

Considering compatibility, the default should be `false`.

TokageItLab reviewed (commented):
The implementation approach seems correct. I would like to have more consensus on the setting names and how to save editor settings in addition to my own opinion. Maybe @KoBeWi or @Mickeon ?

shadow-foss commented:
Just pushed a small update: changed the default to false for compatibility with the previous behavior (inserting at mouse position).
Also open to renaming the setting,other variables if there's a better consensus on the wording :)

Mickeon commented on doc/classes/EditorSettings.xml:
```diff
@@ -547,6 +547,9 @@
 			Default step mode for [AnimationPlayer] (seconds or FPS). The option is remembered locally for a scene and this option only determines the default value when scene doesn't have local state yet.
 		</member>
+		<member name="editors/animation/insert_at_timeline_cursor" type="bool" setter="" getter="">
+			If [code]true[/code], animation keys and markers are inserted at the timeline cursor position. If [code]false[/code], they are inserted at the mouse position.
```
Maybe...?
```suggestion
			If [code]true[/code], animation keys and markers are inserted at the current time in the animation. If [code]false[/code], they are inserted at the mouse cursor's position.
```

TokageItLab reviewed (changes requested):
Oh sorry, I overlooked, but we need to use double for the time.

shadow-foss commented on editor/animation_track_editor.cpp:
```diff
@@ -4931,6 +4933,18 @@ bool AnimationTrackEditor::is_snap_keys_enabled() const {
+	return insert_at_current_time->is_pressed();
+}
+
+float AnimationTrackEditor::get_insert_at_current_time_position_if_enabled(float p_fallback_ofs) const {
+	if (is_insert_at_current_time_enabled()) {
+		return timeline->get_play_position();
+	} else {
+		return p_fallback_ofs;
+	}
+}
+
```
```animation_track_editor.cpp
void AnimationTrackEditor::_insert_key_from_track(float p_ofs, int p_track) {
    ERR_FAIL_INDEX(p_track, animation->get_track_count());
    if (snap_keys->is_pressed() && step->get_value() != 0) {
        p_ofs = snap_time(p_ofs);
    }
    p_ofs = get_insert_at_current_time_position_if_enabled(p_ofs); 
    // ... rest of function
    
the function is used in contexts where the offset parameter is already `float`, the insert_at_current_time in func name is from the button name and has nothing to do with time.
i will change it too use reference passing for fallback

shadow-foss replied:
@TokageItLab Since the function name `get_insert_at_current_time_position_if_enabled()` might be misleading (as it returns an position, not a time), what do you think about renaming it to `resolve_insertion_offset(float &r_offset)` ?

If that sounds reasonable, I’ll go ahead and rename it and refactor to use reference passing for fallbacks. Let me know your thoughts!

KoBeWi commented:
I think it would be nice if it also affected duplicate operation. I remember the AnimationPlayer used to insert duplicated keys at current time, but it was changed at some point.

I also noticed the toolbar in animation player has became quite crowded 🤔 
![image](https://github.com/user-attachments/assets/c1daa887-5758-4f79-83f1-91dcda608bb5)
And this adds yet another button.

TokageItLab commented:
> Skip the offset adjustment when the toggle is enabled:

Some operations (such as baking) require keys to be inserted, but I assume this was a hack to prevent the index from being corrupted due to overwriting (deletion).

In the case of markers, the current skip operation is okay, but in the case of keys, it will be necessary to test whether there are any problems.

shadow-foss commented:
>I think it would be nice if it also affected duplicate operation. I remember the AnimationPlayer used to insert duplicated keys at current time, but it was changed at some point.

Currently, AnimationPlayer still pastes or duplicates keys at the timeline cursor when using keyboard shortcuts. However, when using the menu, keys are inserted at the mouse position. I tried to make the toggle affect paste/duplicate as well, but both functions currently overwrite existing keys without any checks, which could lead to unintended behavior—as @TokageItLab pointed out.
I'll add support for the toggle to paste and duplicate once I figure out a proper way to handle overwriting of existing keys.

TokageItLab commented:
Existing keys should be always overwritten with the latest pasted/inserted keys.

However, in cases where multiple keys are selected and processed sequencialy, care must be taken not to break the indices.

shadow-foss commented:
I tested pasting and duplicating keys at the same position multiple times. They currently overwrite existing keys as expected, and undo/redo works correctly.
Since this PR doesn't change that logic, and insert still avoids collisions via offset adjustment, I think it's safe to leave as-is.
However, for long-term consistency and clarity, I’d be happy to open a follow-up issue discussing whether paste/duplicate should also warn or prevent overwrites like insert does.
Also wanted to share my changes here before pushing it.

made the toggle button affect paste and duplicate operations:
```diff
-           emit_signal(SNAME("duplicate_request"), insert_at_pos, true);
+          emit_signal(SNAME("duplicate_request"), insert_at_pos, !editor->is_insert_at_current_time_enabled());
-           emit_signal(SNAME("paste_request"), insert_at_pos, true);
+          emit_signal(SNAME("paste_request"), insert_at_pos, !editor->is_insert_at_current_time_enabled());
```
changed the function resolve_insertion_offset() to use reference passing and changed function call accord to it:
```diff
-float AnimationTrackEditor::get_insert_at_current_time_position_if_enabled(float p_fallback_ofs) const {
+void AnimationTrackEditor::resolve_insertion_offset(float &r_offset) const {
        if (is_insert_at_current_time_enabled()) {
-               return timeline->get_play_position();
-       } else {
-               return p_fallback_ofs;
+               r_offset = timeline->get_play_position();
        }
 }
```
Let me know if this looks good or if anything should be tweaked before I push it.

TokageItLab commented:
Since we need to [squash](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase) them and make a single commit in the end, you can keep pushing them so that we can build them :)

shadow-foss commented:
Everything requested so far has been addressed and pushed — the toggle now affects insert, paste, and duplicate as expected.
I also refactored the resolve_insertion_offset() function to use reference passing as discussed, and kept it float since it's used in float contexts.
No further changes planned in this PR. If needed, I can open a follow-up issue to improve the insert logic and overwrite handling.
Would be great if someone could take a final look at the code. Thanks again for all the feedback :)

KoBeWi reviewed (approved):
Looks good.
Still needs animation team review probably.

shadow-foss commented:
Accidently, closed the pr while using mobile client. mb

TokageItLab reviewed (approved):
Tested and LGTM, please squash commits into one to follow our [PR workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase).

shadow-foss commented:
squashed all the commits to a single commit :)

TokageItLab commented:
CI test seems to be malfunctioning, so we will run the job again later. I think the code is fine now, thanks.

shadow-foss commented:
Ran `svgo` again on the icon file since the CL checks were failing due to it. It reduced the size slightly (~0.02 KB). I've force-pushed the updated commit

TokageItLab commented:
If there is a problem with the svg itself, the change suggestion should be displayed as an error, so I suspect that the link to the svgo module on the CI side is broken. I think the svg is fine as it is.

TokageItLab commented on editor/icons/InsertAtCurrentTime.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="16" height="16"><path fill="#FFF" d="M5 1 1 5l4 4 4-4zm11 10.46L9.857 8.934l-1.571-.648.648 1.571L11.46 16l.913-2.72 1.815 1.818.91-.91-1.819-1.815z"/></svg>
```
```suggestion
<svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" width="16" height="16"><path fill="#FFF" d="M5 1 1 5l4 4 4-4zm11 10.46L9.857 8.934l-1.571-.648.648 1.571L11.46 16l.913-2.72 1.815 1.818.91-.91-1.819-1.815z"/></svg>

```

shadow-foss commented:
added a newline at the end of svg file, the cl tests would work now ig.
